File: app/src/main/java/de/andicodes/vergissnix/data/TaskDao.java
package de.andicodes.vergissnix.data;

import java.time.ZonedDateTime;
import java.util.List;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

@Dao
public abstract class TaskDao {

    @Query("select * from Task")
    public abstract LiveData<List<Task>> allTasks();

    @Query("select * from Task where id = :id")
    public abstract Task getTask(long id);

    public void saveTask(Task task) {
        if (task.getTimeCreated() == null) {
            task.setTimeCreated(ZonedDateTime.now());
        }
        insertTask(task);
    }

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract void insertTask(Task task);

    @Delete
    public abstract void deleteTask(Task task);
}

File: app/src/main/java/de/andicodes/vergissnix/data/TaskDao.kt
package de.andicodes.vergissnix.data

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import kotlinx.coroutines.flow.Flow
import java.time.ZonedDateTime

@Dao
interface TaskDao {
    @Query("select * from Task where timeDone is null order by time")
    fun allTasks(): Flow<List<Task>>

    @Query("select * from Task where timeDone is not null order by time")
    fun doneTasks(): Flow<List<Task>>

    @Query("select * from Task where timeDone is null and (time is null or time <= :until) order by time")
    fun allTasks(until: ZonedDateTime): Flow<List<Task>>

    @Query("select * from Task where id = :id")
    suspend fun getTask(id: Long): Task?

    /**
     * Save a task in the database.
     *
     * @param task the task to save
     * @return the same task, but updated with the id after the insertion
     */
    suspend fun saveTask(task: Task): Task {
        if (task.timeCreated == null) {
            task.timeCreated = ZonedDateTime.now()
        }
        val id = insertTask(task)
        task.id = id
        return task
    }

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertTask(task: Task): Long

    @Delete
    suspend fun deleteTask(task: Task)
}